tweets/2022-04-27/2022-04-27-tweets.md

```diff
@@ -0,0 +1,342 @@
+  };
+
+  constructor() {
+    super();
+    this.msg = ' ';
+    this.counter = 0;
+  }
+
+  updated(changedProperties) {
+    super.updated(changedProperties);
+    if (i < fullText.length) {
+      setTimeout(() => {
+        this.msg += fullText[i];
+        i += 1;
+      }, Math.floor(Math.random() * 50) + 40);
+    }
+  }
+
+  render() {
+    return html`
+      <p>🤔 Hello <span>${this.msg}</span>${'🤯'.repeat(this.counter)}</p>
+      <button @click=${this._inc}>+</button>
+    `;
+  }
+
+  _inc() {
+    if (i >= fullText.length) {
+      this.counter += 1;
+    }
+  }
+
+  static styles = [
+    css`
+      button {
+        font-size: 200%;
+        width: 64px;
+        height: 64px;
+        border: none;
+        border-radius: 10px;
+        background-color: seagreen;
+        color: white;
+      }
+    `,
+  ];
+}
+```
+
+---
+
+Now if we add this to our page we get...
+
+✅ it rendered
+❌ no typing effect
+❌ button is not doing anything
+
+```js
+export async function registerCustomElements() {
+  // server-only components
+  customElements.define('hello-wave', await import('#components/HelloWave').then(m => m.HelloWave));
+  customElements.define('hello-typer', await import('#components/HelloTyper').then(m => m.HelloTyper));
+}
+
+export default () => html`
+  <h1>Hello World</h1>
+  <hello-wave></hello-wave>
+  <hello-typer></hello-typer>
+`;
+```
+
+![Typer Static](./typer-static.png)
+
+---
+
+We get this because our component is now more then "just" text.
+It now has interactive functionality.
+
+Which means that if we only sent HTML & CSS to the browser it will not be enough.
+
+---
+
+We can now decide to skip the server render and render the component purely client side.
+
+➡️ "nothing" and then the component gets rendered
+➡️ brings a layout shift
+➡️ delays initial render
+➡️ requires search engines to execute JS to see the content (some do, some don't)
+
+```js
+export async function registerCustomElements() {
+  // server-only components
+  customElements.define('hello-wave', await import('#components/HelloWave').then(m => m.HelloWave));
+  // client-only components
+  // 'hello-typer': () => import('#components/HelloTyper').then(m => m.HelloTyper),
+}
+
+export default () => html`
+  <h1>Hello World</h1>
+  <hello-wave></hello-wave>
+  <hello-typer loading="client"></hello-typer>
+`;
+```
+
+<video src="./typer-client.mp4"></video>
+
+---
+
+Or we do ✨ Progressive Hydration ✨
+
+Progressive stands for?
+
+➡️ Start with 0 JavaScript of components
+➡️ Load components as needed
+
+---
+
+Hydrations stands for?
+
+1️⃣ Server render (ship HTML with Declarative Shadow Dom)
+2️⃣ Client load the components JS Code
+3️⃣ Rerender the component (sync DOM/JS - no change)
+4️⃣ Now the component can be interacted with
+
+---
+
+By hydrating our component as soon as it becomes visible we get
+
+➡️ component gets rendered as static content
+➡️ no layout shift or delayed render
+➡️ interactive component as loading conditions are met
```

Comment:
Again with bullet points starting in "mid sentence", we could format them a bit different

By hydrating our component as soon as it becomes visible we get

➡️ component getting rendered as static content
➡️ no layout shift or delayed render
➡️ interactive components as loading conditions are met